Objects/regions.c

```diff
@@ -92,9 +116,324 @@ struct regionmetadata {
+    (regionmetadata_inc_osc(_Py_CAST(Py_region_ptr_t, self)))
+
+static void regionmetadata_dec_osc(Py_region_ptr_t self_raw)
+{
+    self_raw = regionmetadata_get_root(self_raw);
+
+    // Constant regions don't keep track of open subregions.
+    if (IS_CONST_REGION(self_raw)) {
+        return;
+    }
+
+    _Py_CAST(regionmetadata*, self_raw)->osc -= 1;
+}
+#define regionmetadata_dec_osc(self) \
+    (regionmetadata_dec_osc(_Py_CAST(Py_region_ptr_t, self)))
+
+static void regionmetadata_set_parent(regionmetadata* self, regionmetadata* parent) {
+    self = regionmetadata_get_root_or_null(self);
+    if (!self) {
+        assert(false && "Can't set the parent on a constant region");
+        return;
+    }
+
+    Py_region_ptr_t old_parent = Py_region_ptr(self->parent);
+    regionmetadata_dec_rc(old_parent);
+    if (regionmetadata_is_open(self)) {
+        regionmetadata_dec_osc(old_parent);
+    }
+
+    Py_region_ptr_t new_parent = regionmetadata_get_root(parent);
+    assert(!IS_CONST_REGION(new_parent) && "Constant regions can't be used as a parent");
+    self->parent = Py_region_ptr_with_tags(new_parent);
+
+    regionmetadata_inc_rc(new_parent);
+    if (regionmetadata_is_open(self)) {
+        regionmetadata_inc_osc(new_parent);
+    }
+}
+
+static regionmetadata* regionmetadata_get_parent(regionmetadata* self) {
+    self = regionmetadata_get_root_or_null(self);
+    if (!self) {
+        // Constant regions never have a parent
+        return NULL;
+    }
+
+    // Make sure we actually look at the real parent and not a region that
+    // we're merged into.
+    Py_region_ptr_t parent_raw = regionmetadata_get_root(Py_region_ptr(self->parent));
+    assert(!IS_CONST_REGION(parent_raw) && "Constant regions can never be a parent.");
+    regionmetadata* parent = _Py_CAST(regionmetadata*, parent_raw);
+
+    // Here we set the parent to the representative we received.
+    // In most cases this will do nothing, but in others it can decrese
+    // future lookup times.
+    regionmetadata_set_parent(self, parent);
+
+    return _Py_CAST(regionmetadata*, parent);
+}
+#define regionmetadata_get_parent(self) \
+    regionmetadata_get_parent(_Py_CAST(regionmetadata*, self))
+
+static bool regionmetadata_has_parent(regionmetadata* self) {
+    return regionmetadata_get_parent(self) != NULL;
+}
+
+// This implementation merges `self` into `other`. Merging is not allowed
+// to break external uniqueness. It's therefore not allowed for both regions
+// to have a parent. Except cases, where one region has the other region as
+// it's parent.
+//
+// This function expects `self` to be a valid object.
+__attribute__((unused))
+static PyObject* regionmetadata_merge(regionmetadata* self, Py_region_ptr_t other) {
+    self = regionmetadata_get_root_or_null(self);
+    other = regionmetadata_get_root(other);
+    if (!self) {
+        assert(false && "There should never be an attempt to merge constant regions");
```

Comment:
When I read this I first thought you don't support merging into to locals, but I now realise this sentence is just talking about the "source" of a merge, not both "source and target".